internal/terraform/workspace/workspace_manager.go

```diff
@@ -0,0 +1,161 @@
+package workspace
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-ls/internal/terraform/discovery"
+	"github.com/hashicorp/terraform-ls/internal/terraform/exec"
+	"github.com/hashicorp/terraform-ls/internal/terraform/lang"
+	"github.com/hashicorp/terraform-ls/internal/terraform/schema"
+)
+
+type workspaceManager struct {
+	ctx           context.Context
+	ws            map[string]*workspace
+	tfExecPath    string
+	tfExecTimeout time.Duration
+	tfExecLogPath string
+	logger        *log.Logger
+
+	newWorkspace WorkspaceFactory
+}
+
+type WorkspaceFactory func(context.Context, string) (*workspace, error)
+
+type WorkspaceManagerFactory func(context.Context) WorkspaceManager
+
+func NewWorkspaceManager(ctx context.Context) WorkspaceManager {
+	return newWorkspaceManager(ctx)
+}
+
+func newWorkspaceManager(ctx context.Context) *workspaceManager {
+	wm := &workspaceManager{
+		ctx:    ctx,
+		ws:     make(map[string]*workspace, 0),
+		logger: defaultLogger,
+	}
+	wm.newWorkspace = wm.defaultWorkspaceFactory
+	return wm
+}
+
+func (wc *workspaceManager) defaultWorkspaceFactory(ctx context.Context, dir string) (*workspace, error) {
+	w := newWorkspace(ctx)
+
+	w.SetLogger(wc.logger)
+
+	d := &discovery.Discovery{}
+	w.tfDiscoFunc = d.LookPath
+	w.tfNewExecutor = exec.NewExecutor
+	w.newSchemaStorage = schema.NewStorage
+
+	w.tfExecPath = wc.tfExecPath
+	w.tfExecTimeout = wc.tfExecTimeout
+	w.tfExecLogPath = wc.tfExecLogPath
+
+	return w, w.init(ctx, dir)
+}
+
+func (wc *workspaceManager) SetTerraformExecPath(path string) {
+	wc.tfExecPath = path
+}
+
+func (wc *workspaceManager) SetTerraformExecLogPath(logPath string) {
+	wc.tfExecLogPath = logPath
+}
+
+func (wc *workspaceManager) SetTerraformExecTimeout(timeout time.Duration) {
+	wc.tfExecTimeout = timeout
+}
+
+func (wc *workspaceManager) SetLogger(logger *log.Logger) {
+	wc.logger = logger
+}
+
+func (wc *workspaceManager) AddWorkspace(dir string) error {
+	_, exists := wc.ws[dir]
+	if exists {
+		return fmt.Errorf("workspace %s was already added", dir)
+	}
+
+	w, err := wc.newWorkspace(wc.ctx, dir)
+	if err != nil {
+		return err
+	}
+
+	wc.ws[dir] = w
+
+	return nil
+}
+
+func (wc *workspaceManager) WorkspaceByPath(path string) (Workspace, error) {
+	if w, ok := wc.ws[path]; ok {
```

Comment:
You may want to normalize the keys with a `filepath.ToSlash` or something here.

Reply:
🤔   I don't disagree, but also I wonder when would that matter? Are you thinking of https://github.com/hashicorp/terraform-ls/issues/75 or https://github.com/hashicorp/terraform-ls/issues/74 ?

Reply:
It turns out module paths are [always persisted with Unix-style separators](https://github.com/hashicorp/terraform/blob/bc386234d528747ad244574f72f1fc30a486d5c2/internal/modsdir/manifest.go#L131-L134) so good catch - this probably wouldn't work on Windows! 🙈

Reply:
I added some logic to account for module paths being stored with forward slashes, but I think I'd rather avoid converting it back and forth when comparing the paths, for now.

That said I'm not opposed to implementing it later - that is also why the `map[string]*workspace` is an implementation detail, rather than something directly exposed.